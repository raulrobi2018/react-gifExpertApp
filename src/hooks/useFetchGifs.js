import {useState} from "react";

/* Los custom hooks funcionan como si fueran functional components. Pueden utilizar muchas caracterÃ­sticas
de react y son simplemente funciones */

export const useFetchGifs = () => {
    const [state, setstate] = useState({
        data: [],
        loading: true
    });

    setTimeout(() => {
        setstate({
            data: [1, 2, 3, 4, 5, 6],
            loading: false
        });
    }, 3000);

    return state; //retorna data = [] y loading = true
};
