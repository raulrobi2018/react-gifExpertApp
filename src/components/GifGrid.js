import {useFetchGifs} from "../hooks/useFetchGifs";
import GifGridItem from "./GifGridItem";
import PropTypes from "prop-types";

export const GifGrid = ({category}) => {
    const {data: images, loading} = useFetchGifs(category);

    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{category}</h3>

            {loading && (
                <p className="animate__animated animate__flash">Cargando...</p>
            )}

            <div className="cardGrid">
                {images.map((img) => {
                    // Al enviar '...img' estoy enviando cada una de las propiedades de la imagen
                    // de manera independiente
                    return <GifGridItem {...img} key={img.id} />;

                    // Otra forma
                    // return <GifGridItem img={img} key={img.id} />;
                })}
            </div>
        </>
    );
};

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
};

export default GifGrid;
