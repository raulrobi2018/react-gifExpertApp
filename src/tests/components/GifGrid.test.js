import {shallow} from "enzyme";
import GifGrid from "../../components/GifGrid";
import React from "react";
import {useFetchGifs} from "../../hooks/useFetchGifs";

//Simula cualquier llamada al useFetchGifs
jest.mock("../../hooks/useFetchGifs");

describe("Testing GifGrid", () => {
    const categ = "Cats";

    test("should return the GifGrid correctly", () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        const wrapper = shallow(<GifGrid category={categ} />);
        expect(wrapper).toMatchSnapshot();
    });

    test("should display the items when images are loaded with useFetchGifs", () => {
        const gifs = [
            {
                id: "ABC",
                url: "http://localhost/cualquiercosa",
                title: "Cualquier cosa"
            }
        ];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={categ} />);
        expect(wrapper).toMatchSnapshot();
    });
});
