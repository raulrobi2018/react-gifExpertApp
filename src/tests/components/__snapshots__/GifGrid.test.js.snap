// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GifGrid should display the items when images are loaded with useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Cats
  </h3>
  <div
    className="cardGrid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="http://localhost/cualquiercosa"
    />
  </div>
</Fragment>
`;

exports[`Testing GifGrid should return the GifGrid correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Cats
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Cargando...
  </p>
  <div
    className="cardGrid"
  />
</Fragment>
`;
